File: epi_judge_python/search_for_min_max_in_array.py
import collections
from sys import exit

from test_framework import generic_test, test_utils
from test_framework.test_failure import PropertyName

MinMax = collections.namedtuple('MinMax', ('smallest', 'largest'))


def find_min_max(A):
    # Implement this placeholder.
    return MinMax(0, 0)


def res_printer(prop, value):
    def fmt(x):
        return 'min: {}, max: {}'.format(x[0], x[1]) if x else None

    if prop in (PropertyName.EXPECTED, PropertyName.RESULT):
        return fmt(value)
    else:
        return value


if __name__ == '__main__':
    exit(
        generic_test.generic_test_main(
            'search_for_min_max_in_array.tsv',
            find_min_max,
            res_printer=res_printer))

File: epi_judge_cpp/search_for_min_max_in_array.cc
#include <vector>

#include "test_framework/generic_test.h"
#include "test_framework/serialization_traits.h"
using std::vector;
struct MinMax {
  int smallest, largest;
};

MinMax FindMinMax(const vector<int>& A) {
    if (A.size() <= 1)
        return { A.front(), A.front() };
    
    std::pair<int, int> global_min_max = std::minmax(A[0], A[1]);
    for (int i = 2; i < A.size() - 1; i++) {
        std::pair<int, int> local_min_max = std::minmax(A[i], A[i + 1]);

        global_min_max = {  std::min(global_min_max.first, local_min_max.first), 
                            std::max(global_min_max.second, local_min_max.second) };
    }
    if (A.size() % 2) {
       global_min_max = { std::min(global_min_max.first, A.back()),
                           std::max(global_min_max.second, A.back()) };
    }
    
    return { global_min_max.first, global_min_max.second};
}
namespace test_framework {
template <>
struct SerializationTrait<MinMax> : UserSerTrait<MinMax, int, int> {};
}  // namespace test_framework

bool operator==(const MinMax& lhs, const MinMax& rhs) {
  return std::tie(lhs.smallest, lhs.largest) ==
         std::tie(rhs.smallest, rhs.largest);
}

std::ostream& operator<<(std::ostream& out, const MinMax& x) {
  return out << "min: " << x.smallest << ", max: " << x.largest;
}

int main(int argc, char* argv[]) {
  std::vector<std::string> args{argv + 1, argv + argc};
  std::vector<std::string> param_names{"A"};
  return GenericTestMain(args, "search_for_min_max_in_array.cc",
                         "search_for_min_max_in_array.tsv", &FindMinMax,
                         DefaultComparator{}, param_names);
}
